shell/browser/api/electron_api_push_notifications_mac.mm

```diff
@@ -0,0 +1,64 @@
+// Copyright (c) 2022 Asana, Inc.
+// Use of this source code is governed by the MIT license that can be
+// found in the LICENSE file.
+
+#include "shell/browser/api/electron_api_push_notifications.h"
+
+#include <string>
+
+#include <utility>
+#include <vector>
+#import "shell/browser/mac/electron_application.h"
+#include "shell/common/gin_converters/value_converter.h"
+#include "shell/common/gin_helper/promise.h"
+
+namespace electron {
+
+namespace api {
+
+// This set maintains all the promises that should be fulfilled
+// once macOS registers, or fails to register, for APNS
+std::vector<gin_helper::Promise<std::string>> apns_promise_set_;
+
+v8::Local<v8::Promise> PushNotifications::RegisterForAPNSNotifications(
+    v8::Isolate* isolate) {
+  gin_helper::Promise<std::string> promise(isolate);
+  v8::Local<v8::Promise> handle = promise.GetHandle();
+
+  [[AtomApplication sharedApplication]
+      registerForRemoteNotificationTypes:NSRemoteNotificationTypeBadge |
+                                         NSRemoteNotificationTypeAlert |
+                                         NSRemoteNotificationTypeSound];
+
+  apns_promise_set_.emplace_back(std::move(promise));
+  return handle;
+}
+
+void PushNotifications::ResolveAPNSPromiseSetWithToken(
+    const std::string& token_string) {
+  for (auto& promise : apns_promise_set_) {
+    promise.Resolve(token_string);
+  }
+  apns_promise_set_.clear();
```

Comment:
Mega edge-case nitpick here, but to be on the safe side, it's best to make sure lists aren't modified during iteration. As a bonus, we can simplify this code a little:

```suggestion
  std::vector<gin_helper::Promise<std::string>> promises = std::move(apns_promise_set_);
  for (auto& promise : promises)
    promise.Resolve(token_string);
```

`std::move` on a vector clears out the original vector. If `register` is called again while we're in the middle of resolving promises, it won't affect the iteration, since we're operating on a separate vector now.

Same for the error method below :)